include/navigation/differencers.h

```diff
@@ -149,36 +158,45 @@ class RMSDiff
  */
 struct CorrCoefficient {
     /*
-     * NB: This template parameter is unused, but left in for consistency with
-     * other differencer classes.
+     * NB: We don't need any scratch storage, but the user can choose to have
+     * some (e.g. for the HOG flavour of perfect memory).
      */
-    template<class VecType = void>
+    template<class VecType = std::tuple<>>
     class Internal
+      : public DifferencerBase<CorrCoefficient::Internal<VecType>, VecType>
     {
     public:
         float operator()(cv::InputArray &src1, cv::InputArray &src2,
                          const ImgProc::Mask &mask1 = {},
                          const ImgProc::Mask &mask2 = {})
         {
-            // Don't support masks for now
-            BOB_ASSERT(mask1.empty() && mask2.empty());
-
-            const auto allSame = [](cv::InputArray &arr) {
-                const auto m = arr.getMat();
-                BOB_ASSERT(m.type() == CV_8UC1);
-                auto beg = m.datastart;
-                auto end = m.dataend;
-                return std::all_of(beg + 1, end,
-                                   [beg](uint8_t val) { return val == *beg; });
-            };
-            if (allSame(src1) || allSame(src2)) {
+            BOB_ASSERT(src1.type() == src2.type());
+
+            /*
+             * NB: We could easily support other mat types but these are the
+             * only ones we care about for now.
+             */
+            bool nonUniqueInputArray;
+            switch (src1.type()) {
+            case CV_8UC1:
+                nonUniqueInputArray = allSame<uint8_t>(src1) || allSame<uint8_t>(src2);
+                break;
+            case CV_32FC1:
```

Comment:
is this to handle HOG features? I seem to remember raw images are always ``CV_8UC1``

Reply:
Yes it is